sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -740,8 +751,10 @@ object NullPropagation extends Rule[LogicalPlan] {
     case _ => false
   }
 
-  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case q: LogicalPlan => q transformExpressionsUp {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsAnyPattern(NULL_CHECK, NULL_LITERAL, COUNT, CAST), ruleId) {
```

Comment:
Why do we need `CAST` here?

Reply:
I tried to capture the first case:
 case e @ WindowExpression(Cast(Literal(0L, _), _, _), _)

I'm not sure why this pattern is related to NullPropagation, but it probably deserve a separate change.
I updated the condition to be more restrictive since CAST may be common in certain queries:

t.containsAnyPattern(NULL_CHECK, NULL_LITERAL, COUNT)
        || t.containsAllPatterns(WINDOW_EXPRESSION, CAST, LITERAL)

Reply:
OK. ​I think now developers need to spend extra effort on choosing the Bits here.  And there seems no good solution for avoiding regressions except for asking developers to add more unit tests.
Do you have any idea to improve the framework for this?

Reply:
>> seems no good solution for avoiding regressions except for asking developers to add more unit tests

- For an existing rule that we're adding this pruning, yes, we need to be careful, particularly when a rule has complex case patterns, although many rules are simple and intuitive.

- For a new rule or a new case branch in an existing rule, it seems that the regression risk doesn't change much? if a case pattern branch is not covered by unit tests, there's also a regression risk even without bits pruning, e.g., the case pattern could be wrong too.